Fix generated ESM, revamp build system

hobbes7878 opened the pull request:
Attempt to close #111. 

Done:
- [x] Updates build system using [@sveltejs/package](https://www.npmjs.com/package/@sveltejs/package).
- [x] Makes package an ES module.
- [x] Converts all the scripts to typescript and fixes any outstanding type issues.
- [x] Updates a buncha dev dependencies. 
- [x] Working with storybook 7.1, updated to use Vite framework.
- [x] Fix/upgrade tests (though, drops snapshot tests in favour of using Chromatic)

<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>3.0.5--canary.113.4269195.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/addon-svelte-csf@3.0.5--canary.113.4269195.0
  # or 
  yarn add @storybook/addon-svelte-csf@3.0.5--canary.113.4269195.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

JReinhold commented:
This is looking great @hobbes7878 ! 🥰

It's going to be interesting to see if this will still work in a Webpack-based project, but if not, it would be okay to drop support for that, we're seeing very little usage of Storybook 7 + Webpack + Svelte + this addon.

hobbes7878 commented:
Thanks @JReinhold. Happy if it helps.

On tests, moving the whole package to ECMAScript modules has made it a little tricky to work with Jest. Right now, I've replaced it with vitest for all the unit tests, which was easier to get going. 

That leaves the snapshot tests. I wonder if Chromatic is a good enough replacement for the storyshots. I've updated this branch with that in mind for now, but happy if someone wants to give me any hints/examples how to bring those back...

hobbes7878 commented:
> It's going to be interesting to see if this will still work in a Webpack-based project, but if not, it would be okay to drop support for that, we're seeing very little usage of Storybook 7 + Webpack + Svelte + this addon.

Yeah, I'm not sure, but there weren't any webpack-specific plugins in it originally, so as long as webpack can tolerate the move to ESM and this did _previously_ work there, it _should_ be fine??

davemackintosh commented:
I've just installed this and I can confirm the issue is fixed for me. Thank you! 🙏

davemackintosh commented:
I've just installed this and I can confirm the issue is fixed for me. Thank you! 🙏

olafurw commented:
Tested this also on a branch that had this issue. Fixes the issue for me.

ndelangen commented on package.json:
```diff
@@ -11,9 +11,21 @@
     "url": "https://github.com/storybookjs/addon-svelte-csf"
   },
   "license": "MIT",
-  "main": "dist/cjs/index.js",
-  "module": "dist/esm/index.js",
-  "browser": "dist/cjs/index.js",
```
You say this works in storybook?
I would expect this to be broken, but honestly I've not yet tried, so maybe I'm wrong.

What i know of the preset-loading mechanism in storybook is what when an addon-entry is added, storybook will try resolving `${name}/preset` (with node).
It will do the same for `manager` and `preview` FYI.

Maybe with `type: module` these exports fields are actually respected/used?

hobbes7878 replied:
Yeah, `type: "module"` is the key there as far as I understand it, which is mostly from [sindresorhus's explanation](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#faq)...

But yes, working in Node 16+ with Storybook 7.1+ in the repo.

ndelangen commented on src/plugins/vite-svelte-csf.ts:
```diff
@@ -7,7 +7,7 @@ import { fileURLToPath } from 'url';
 import { readFileSync } from 'fs';
 
-const parser = fileURLToPath(new URL('../parser/collect-stories.js', import.meta.url));
+const parser = fileURLToPath(new URL('../parser/collect-stories.js', import.meta.url)).replace(/\\/g, "\\\\");
```
I'd recommend making this very obvious why this is the way it is perhaps with a clear code-comment?, and possibly even ensure this continues working with a unit-test.

hobbes7878 replied:
Added handling across any dynamic imports in https://github.com/storybookjs/addon-svelte-csf/pull/113/commits/fa5505bda90c1c1d06a6a6b3d508b82d6813ef14. 

@ilsubyeega don't know if you'd be able to help with a test?

ilsubyeega replied:
I can say the current branch works on my machine (win11-x64) with `yarn test` and storybook live view.
Also, I'm not sure about creating other test tasks for this.

JReinhold commented on tsconfig.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "extends": "@tsconfig/svelte/tsconfig.json",
+  "compilerOptions": {
```
Missing `strict: true` maybe? Or does that not work for some reason? I see it's explicitly `false` in [`@tsconfig/svelte`](https://www.npmjs.com/package/@tsconfig/svelte)

hobbes7878 replied:
Looks like `strict: true` is fine: https://github.com/sveltejs/kit/issues/4395

Will add.

JReinhold reviewed (changes requested):
Absolutely fantastic job, I also love the way you've cleaned up the code all over the place. The "conversion" from storyshots to Chromatic is also \*chef's kiss\* (and Jest -> vitest), storyshots really shouldn't be used anymore.

I have left some comments in the code, and then a big issue is that the Storybook fails to build, which is evident by [the Chromatic check on this PR failing](https://github.com/storybookjs/addon-svelte-csf/actions/runs/5668946724/job/15360661875), and I was able to reproduce it locally. It's failing with some CJS->ESM shenanigans in `cli-table3`. I was able to fix it locally by upgrading all Storybook dependencies from `7.1.0` to `7.1.1`, could you try that? I'll investigate on the core end why this is.

PS. @hobbes7878 would you like to become a maintainer of this project?

hobbes7878 commented on package.json:
```diff
@@ -22,68 +34,65 @@
     "svelte": "^3.50.0 || ^4.0.0",
     "svelte-loader": "^3.1.2",
-    "vite": "^3.0.0 || ^4.0.0"
+    "vite": "^4.0.0"
```
Widened that range back.

hobbes7878 commented on package.json:
```diff
@@ -22,68 +34,65 @@
-    "@sveltejs/vite-plugin-svelte": "^1.0.0 || ^2.0.0",
+    "@storybook/svelte": "^7.0.0",
+    "@storybook/theming": "^7.0.0",
+    "@sveltejs/vite-plugin-svelte": "^2.0.0",
```
Widened that range back.

hobbes7878 commented:
Thanks @JReinhold. Addressed all those suggestions, I think.

I'm still seeing that `cli-table3` issue with 7.1.1 Storybook [in the Chromatic workflow](https://github.com/storybookjs/addon-svelte-csf/actions/runs/5700009224/job/15449455253?pr=113#step:4:56). Storybook is building fine for me locally. Will keep investigating...

Re: maintainer, I don't mind, sure!

hobbes7878 commented:
> I have left some comments in the code, and then a big issue is that the Storybook fails to build, which is evident by [the Chromatic check on this PR failing](https://github.com/storybookjs/addon-svelte-csf/actions/runs/5668946724/job/15360661875), and I was able to reproduce it locally. It's failing with some CJS->ESM shenanigans in `cli-table3`. I was able to fix it locally by upgrading all Storybook dependencies from `7.1.0` to `7.1.1`, could you try that? I'll investigate on the core end why this is.

Looks like this isn't fixed just by updating storybook to `7.1.1`. But there's something really weird going on with yarn's dependency resolution here for `string-width`.

I can consistently reproduce this issue based on the existence of `yarn.lock`.

So if `yarn.lock` exists and you clear out `node_modules/` and re-`yarn` to install fresh deps from the lockfile, you get the `ERR_REQUIRE_ESM` error in `cli-table3` requiring `string-width`. That's what's of course happening in the workflow, so that busts out.

If I delete `yarn.lock` (or change a dep so it's regenerated) and then `yarn`, no error. There is no difference in the lockfiles in any of these scenarios.

JReinhold commented:
Great work!

That sounds super strange, I have no idea why it behaves that way. Could this just be a bug in Yarn 1? I mean, it hasn't been updated for a while.

Would you be open to migrating to pnpm here and see if that automatically fixes the issue? I don't think there's any deeper integrations with `yarn` at play here, so it should be a simple replacement, but if it's not I'm okay with the current solution you have for now.

hobbes7878 commented:
Not gonna lie, it took a couple tries but migrated to pnpm, which is running fine without any dep overloads. 👍

JReinhold commented on .github/workflows/release.yml:
```diff
@@ -7,22 +7,25 @@ jobs:
           NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
         run: |
-          yarn release
+          npm run release
```
shouldn't all these `npm run` be `pnpm run` instead to use the correct package resolution, or am I misunderstanding something? Same for the lines in `package.json`?

hobbes7878 replied:
Yeah, that was silly of me. Done in [4269195](https://github.com/storybookjs/addon-svelte-csf/pull/113/commits/4269195b43956bc2f7401f480a9012c9ed660581).

JReinhold commented:
Thanks for your hard work and patience @hobbes7878!

oscard0m commented:
It works! Thanks for the amazing job on this @hobbes7878 🌟 
And thanks to the rest of participants as well 
🎖️ @JReinhold 
🎖️ @ilsubyeega
🎖️ @ndelangen
🎖️ @shilman
🎖️ @olafurw
🎖️ @davemackintosh
